WalletWasabi.Gui/MainWindow.xaml.cs

```diff
@@ -78,74 +87,114 @@ private async void MainWindow_ClosingAsync(object? sender, CancelEventArgs e)
-
-				if (Global.WalletManager.AnyCoinJoinInProgress())
+				if (tryToDequeue && Global.WalletManager.AnyCoinJoinInProgress())
 				{
 					var dialog = new CannotCloseDialogViewModel();
 
-					closeApplication = await MainWindowViewModel.Instance.ShowDialogAsync(dialog); // start the deque process with a dialog
+					// Start the deque process with a dialog.
+					closeApplication = await MainWindowViewModel.Instance.ShowDialogAsync(dialog);
 				}
-				else
+
+				if (!closeApplication)
 				{
-					closeApplication = true;
+					//The user aborted the close, let's go to finally.
+					return;
 				}
 
-				if (closeApplication)
+				try
 				{
-					try
-					{
-						if (Global.UiConfig is { }) // UiConfig not yet loaded.
-						{
-							Global.UiConfig.WindowState = WindowState;
+					SaveUiConfig();
 
-							IDocumentTabViewModel? selectedDocument = IoC.Get<IShell>().SelectedDocument;
-							Global.UiConfig.LastActiveTab = selectedDocument is null
-								? nameof(HistoryTabViewModel)
-								: selectedDocument.GetType().Name;
+					Hide();
 
-							Global.UiConfig.ToFile();
-							Logger.LogInfo($"{nameof(Global.UiConfig)} is saved.");
-						}
-
-						Hide();
-						var wm = IoC.Get<IShell>().Documents?.OfType<WalletManagerViewModel>().FirstOrDefault();
-						if (wm is { })
+					if (IoC.Get<IShell>().Documents is { } docs)
+					{
+						foreach (var doc in docs.OfType<WasabiDocumentTabViewModel>().ToArray())
 						{
-							wm.OnClose();
-							Logger.LogInfo($"{nameof(WalletManagerViewModel)} closed.");
+							doc.OnClose();
+							Logger.LogInfo($"ViewModel {doc.Title} was closed.");
 						}
 					}
-					catch (Exception ex)
-					{
-						Logger.LogWarning(ex);
-					}
-
-					Interlocked.Exchange(ref _closingState, 2); //now we can close the app
-					Close(); // start the closing process. Will call MainWindow_ClosingAsync again!
 				}
-				//let's go to finally
+				catch (Exception ex)
+				{
+					Logger.LogWarning(ex);
+				}
+
+				// Now we can close the app.
+				Interlocked.Exchange(ref _closingState, ClosingStateClosed);
+
+				// Don't call the MainWindow_ClosingAsync again.
+				Closing -= MainWindow_ClosingAsync;
+				Close();
 			}
 			catch (Exception ex)
 			{
-				Interlocked.Exchange(ref _closingState, 0); //something happened back to starting point
+				// Something happened back to starting point.
+				Interlocked.Exchange(ref _closingState, ClosingStateNotClosing);
 				Logger.LogWarning(ex);
 			}
 			finally
 			{
-				if (!closeApplication) //we are not closing the application for some reason
+				if (!closeApplication)
 				{
-					Interlocked.Exchange(ref _closingState, 0);
+					// We are not closing the application for some reason.
+					Interlocked.Exchange(ref _closingState, ClosingStateNotClosing);
 					// Re-enable enqueuing coins.
 					Global.WalletManager.SignalQuitPending(false);
 				}
 			}
 		}
+
+		private void SaveUiConfig()
+		{
+			if (Global.UiConfig is { } uiConfig)
```

Comment:
AFAIK, `Global.UiConfig` is always non-null (it's set in `Global`s constructor), so this IF statements seems redundant as `Global` is disposed as the last thing.